In [1]:
#import kaggle
!pip install kaggle
import kaggle


[notice] A new release of pip is available: 24.3.1 -> 26.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [3]:
#download dataset from kaggle
!kaggle datasets download -d hetmengar/ola-and-uber-ride-booking-and-cancellation-data

Dataset URL: https://www.kaggle.com/datasets/hetmengar/ola-and-uber-ride-booking-and-cancellation-data
License(s): CC0-1.0
ola-and-uber-ride-booking-and-cancellation-data.zip: Skipping, found more recently modified local copy (use --force to force download)


In [4]:
#extract from zip file
import zipfile
zip_ref= zipfile.ZipFile('ola-and-uber-ride-booking-and-cancellation-data.zip')
zip_ref.extractall()
zip_ref.close()

In [5]:
import os
os.listdir()

['.ipynb_checkpoints',
 'Bookings.csv',
 'ola-and-uber-ride-booking-and-cancellation-data.zip',
 'Ride_Booking_Cancellation.ipynb']

In [9]:
#import libraries
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt


In [10]:
df=pd.read_csv('Bookings.csv')
df.head(10)

Date      Time     Booking_ID        Booking_Status  \
0  2024-07-26 14:00:00  14:00:00  CNR7153255142    Canceled by Driver   
1  2024-07-25 22:20:00  22:20:00  CNR2940424040               Success   
2  2024-07-30 19:59:00  19:59:00  CNR2982357879               Success   
3   2024-07-22 3:15:00  03:15:00  CNR2395710036  Canceled by Customer   
4   2024-07-02 9:02:00  09:02:00  CNR1797421769               Success   
5   2024-07-13 4:42:00  04:42:00  CNR8787177882               Success   
6   2024-07-23 9:51:00  09:51:00  CNR3612067560               Success   
7  2024-07-11 11:12:00  11:12:00  CNR5374902489    Canceled by Driver   
8  2024-07-01 19:19:00  19:19:00  CNR5030602354      Driver Not Found   
9   2024-07-18 1:31:00  01:31:00  CNR6328453219    Canceled by Driver   

  Customer_ID Vehicle_Type    Pickup_Location  Drop_Location  V_TAT  C_TAT  \
0   CID713523  Prime Sedan        Tumkur Road       RT Nagar    NaN    NaN   
1   CID225428         Bike        Magadi Road        Varthur  203.0   30.0   
2   CID270156    Prime SUV      Sahakar Nagar        Varthur  238.0  130.0   
3   CID581320        eBike         HSR Layout    Vijayanagar    NaN    NaN   
4   CID939555         Mini        Rajajinagar    Chamarajpet  252.0   80.0   
5   CID802429         Mini           Kadugodi    Vijayanagar  231.0   90.0   
6   CID476071         Bike        Tumkur Road     Whitefield  133.0   40.0   
7   CID735691   Prime Plus  Bannerghatta Road  Sarjapur Road    NaN    NaN   
8   CID999840         Mini        Chamarajpet         Peenya    NaN    NaN   
9   CID907133         Auto           RT Nagar        Varthur    NaN    NaN   

   ...      Canceled_Rides_by_Driver Incomplete_Rides Incomplete_Rides_Reason  \
0  ...  Personal & Car related issue              NaN                     NaN   
1  ...                           NaN               No                     NaN   
2  ...                           NaN               No                     NaN   
3  ...                           NaN              NaN                     NaN   
4  ...                           NaN               No                     NaN   
5  ...                           NaN               No                     NaN   
6  ...                           NaN               No                     NaN   
7  ...  Personal & Car related issue              NaN                     NaN   
8  ...                           NaN              NaN                     NaN   
9  ...  Personal & Car related issue              NaN                     NaN   

  Booking_Value  Payment_Method Ride_Distance  Driver_Ratings  \
0           444             NaN             0             NaN   
1           158            Cash            13             4.1   
2           386             UPI            40             4.2   
3           384             NaN             0             NaN   
4           822     Credit Card            45             4.0   
5           173             UPI            41             3.4   
6           140            Cash            49             3.2   
7           344             NaN             0             NaN   
8           839             NaN             0             NaN   
9           893             NaN             0             NaN   

   Customer_Rating  Vehicle Images Unnamed: 20  
0              NaN          #NAME?         NaN  
1              4.0          #NAME?         NaN  
2              4.8          #NAME?         NaN  
3              NaN          #NAME?         NaN  
4              3.0          #NAME?         NaN  
5              4.6          #NAME?         NaN  
6              4.5          #NAME?         NaN  
7              NaN          #NAME?         NaN  
8              NaN          #NAME?         NaN  
9              NaN          #NAME?         NaN  

[10 rows x 21 columns]

In [11]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103024 entries, 0 to 103023
Data columns (total 21 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   Date                        103024 non-null  object 
 1   Time                        103024 non-null  object 
 2   Booking_ID                  103024 non-null  object 
 3   Booking_Status              103024 non-null  object 
 4   Customer_ID                 103024 non-null  object 
 5   Vehicle_Type                103024 non-null  object 
 6   Pickup_Location             103024 non-null  object 
 7   Drop_Location               103024 non-null  object 
 8   V_TAT                       63967 non-null   float64
 9   C_TAT                       63967 non-null   float64
 10  Canceled_Rides_by_Customer  10499 non-null   object 
 11  Canceled_Rides_by_Driver    18434 non-null   object 
 12  Incomplete_Rides            63967 non-null   object 
 13  Incomplete_Rid

In [13]:
df.shape

(103024, 21)

In [17]:
#drop the unnecessary columns
#df=df.drop(columns=['Unnamed: 20'])
#drop the vehicle column also as it does help us in our analysis right now
df=df.drop(columns=['Vehicle Images'])

In [19]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103024 entries, 0 to 103023
Data columns (total 20 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   Date                        103024 non-null  object 
 1   Time                        103024 non-null  object 
 2   Booking_ID                  103024 non-null  object 
 3   Booking_Status              103024 non-null  object 
 4   Customer_ID                 103024 non-null  object 
 5   Vehicle_Type                103024 non-null  object 
 6   Pickup_Location             103024 non-null  object 
 7   Drop_Location               103024 non-null  object 
 8   V_TAT                       63967 non-null   float64
 9   C_TAT                       63967 non-null   float64
 10  Canceled_Rides_by_Customer  10499 non-null   object 
 11  Canceled_Rides_by_Driver    18434 non-null   object 
 12  Incomplete_Rides            63967 non-null   object 
 13  Incomplete_Rid

In [21]:
df.isnull().sum()

Date                               0
Time                               0
Booking_ID                         0
Booking_Status                     0
Customer_ID                        0
Vehicle_Type                       0
Pickup_Location                    0
Drop_Location                      0
V_TAT                          39057
C_TAT                          39057
Canceled_Rides_by_Customer     92525
Canceled_Rides_by_Driver       84590
Incomplete_Rides               39057
Incomplete_Rides_Reason        99098
Booking_Value                      0
Payment_Method                 39057
Ride_Distance                      0
Driver_Ratings                 39057
Customer_Rating                39057
Unnamed: 20                   103024
dtype: int64

In [23]:
df.head()

Date      Time     Booking_ID        Booking_Status  \
0  2024-07-26 14:00:00  14:00:00  CNR7153255142    Canceled by Driver   
1  2024-07-25 22:20:00  22:20:00  CNR2940424040               Success   
2  2024-07-30 19:59:00  19:59:00  CNR2982357879               Success   
3   2024-07-22 3:15:00  03:15:00  CNR2395710036  Canceled by Customer   
4   2024-07-02 9:02:00  09:02:00  CNR1797421769               Success   

  Customer_ID Vehicle_Type Pickup_Location Drop_Location  V_TAT  C_TAT  \
0   CID713523  Prime Sedan     Tumkur Road      RT Nagar    NaN    NaN   
1   CID225428         Bike     Magadi Road       Varthur  203.0   30.0   
2   CID270156    Prime SUV   Sahakar Nagar       Varthur  238.0  130.0   
3   CID581320        eBike      HSR Layout   Vijayanagar    NaN    NaN   
4   CID939555         Mini     Rajajinagar   Chamarajpet  252.0   80.0   

                     Canceled_Rides_by_Customer      Canceled_Rides_by_Driver  \
0                                           NaN  Personal & Car related issue   
1                                           NaN                           NaN   
2                                           NaN                           NaN   
3  Driver is not moving towards pickup location                           NaN   
4                                           NaN                           NaN   

  Incomplete_Rides Incomplete_Rides_Reason  Booking_Value Payment_Method  \
0              NaN                     NaN            444            NaN   
1               No                     NaN            158           Cash   
2               No                     NaN            386            UPI   
3              NaN                     NaN            384            NaN   
4               No                     NaN            822    Credit Card   

   Ride_Distance  Driver_Ratings  Customer_Rating  Unnamed: 20  
0              0             NaN              NaN          NaN  
1             13             4.1              4.0          NaN  
2             40             4.2              4.8          NaN  
3              0             NaN              NaN          NaN  
4             45             4.0              3.0          NaN

In [25]:
df['cancelled_by_customer'] = df['Canceled_Rides_by_Customer'].notna()
df['cancelled_by_driver'] = df['Canceled_Rides_by_Driver'].notna()
#currently we are not interested in reason stated by the drivers or customers so lets convert them into boolean values only

In [27]:
df.head()

Date      Time     Booking_ID        Booking_Status  \
0  2024-07-26 14:00:00  14:00:00  CNR7153255142    Canceled by Driver   
1  2024-07-25 22:20:00  22:20:00  CNR2940424040               Success   
2  2024-07-30 19:59:00  19:59:00  CNR2982357879               Success   
3   2024-07-22 3:15:00  03:15:00  CNR2395710036  Canceled by Customer   
4   2024-07-02 9:02:00  09:02:00  CNR1797421769               Success   

  Customer_ID Vehicle_Type Pickup_Location Drop_Location  V_TAT  C_TAT  ...  \
0   CID713523  Prime Sedan     Tumkur Road      RT Nagar    NaN    NaN  ...   
1   CID225428         Bike     Magadi Road       Varthur  203.0   30.0  ...   
2   CID270156    Prime SUV   Sahakar Nagar       Varthur  238.0  130.0  ...   
3   CID581320        eBike      HSR Layout   Vijayanagar    NaN    NaN  ...   
4   CID939555         Mini     Rajajinagar   Chamarajpet  252.0   80.0  ...   

  Incomplete_Rides Incomplete_Rides_Reason Booking_Value Payment_Method  \
0              NaN                     NaN           444            NaN   
1               No                     NaN           158           Cash   
2               No                     NaN           386            UPI   
3              NaN                     NaN           384            NaN   
4               No                     NaN           822    Credit Card   

   Ride_Distance Driver_Ratings  Customer_Rating  Unnamed: 20  \
0              0            NaN              NaN          NaN   
1             13            4.1              4.0          NaN   
2             40            4.2              4.8          NaN   
3              0            NaN              NaN          NaN   
4             45            4.0              3.0          NaN   

   cancelled_by_customer  cancelled_by_driver  
0                  False                 True  
1                  False                False  
2                  False                False  
3                   True                False  
4                  False                False  

[5 rows x 22 columns]

In [29]:
df = df.drop(columns=[
    'Canceled_Rides_by_Customer',
    'Canceled_Rides_by_Driver'
])


In [31]:
df.head()

Date      Time     Booking_ID        Booking_Status  \
0  2024-07-26 14:00:00  14:00:00  CNR7153255142    Canceled by Driver   
1  2024-07-25 22:20:00  22:20:00  CNR2940424040               Success   
2  2024-07-30 19:59:00  19:59:00  CNR2982357879               Success   
3   2024-07-22 3:15:00  03:15:00  CNR2395710036  Canceled by Customer   
4   2024-07-02 9:02:00  09:02:00  CNR1797421769               Success   

  Customer_ID Vehicle_Type Pickup_Location Drop_Location  V_TAT  C_TAT  \
0   CID713523  Prime Sedan     Tumkur Road      RT Nagar    NaN    NaN   
1   CID225428         Bike     Magadi Road       Varthur  203.0   30.0   
2   CID270156    Prime SUV   Sahakar Nagar       Varthur  238.0  130.0   
3   CID581320        eBike      HSR Layout   Vijayanagar    NaN    NaN   
4   CID939555         Mini     Rajajinagar   Chamarajpet  252.0   80.0   

  Incomplete_Rides Incomplete_Rides_Reason  Booking_Value Payment_Method  \
0              NaN                     NaN            444            NaN   
1               No                     NaN            158           Cash   
2               No                     NaN            386            UPI   
3              NaN                     NaN            384            NaN   
4               No                     NaN            822    Credit Card   

   Ride_Distance  Driver_Ratings  Customer_Rating  Unnamed: 20  \
0              0             NaN              NaN          NaN   
1             13             4.1              4.0          NaN   
2             40             4.2              4.8          NaN   
3              0             NaN              NaN          NaN   
4             45             4.0              3.0          NaN   

   cancelled_by_customer  cancelled_by_driver  
0                  False                 True  
1                  False                False  
2                  False                False  
3                   True                False  
4                  False                False

In [33]:
df['is_incomplete'] = df['Incomplete_Rides'].notna()
df.head()

Date      Time     Booking_ID        Booking_Status  \
0  2024-07-26 14:00:00  14:00:00  CNR7153255142    Canceled by Driver   
1  2024-07-25 22:20:00  22:20:00  CNR2940424040               Success   
2  2024-07-30 19:59:00  19:59:00  CNR2982357879               Success   
3   2024-07-22 3:15:00  03:15:00  CNR2395710036  Canceled by Customer   
4   2024-07-02 9:02:00  09:02:00  CNR1797421769               Success   

  Customer_ID Vehicle_Type Pickup_Location Drop_Location  V_TAT  C_TAT  ...  \
0   CID713523  Prime Sedan     Tumkur Road      RT Nagar    NaN    NaN  ...   
1   CID225428         Bike     Magadi Road       Varthur  203.0   30.0  ...   
2   CID270156    Prime SUV   Sahakar Nagar       Varthur  238.0  130.0  ...   
3   CID581320        eBike      HSR Layout   Vijayanagar    NaN    NaN  ...   
4   CID939555         Mini     Rajajinagar   Chamarajpet  252.0   80.0  ...   

  Incomplete_Rides_Reason Booking_Value  Payment_Method Ride_Distance  \
0                     NaN           444             NaN             0   
1                     NaN           158            Cash            13   
2                     NaN           386             UPI            40   
3                     NaN           384             NaN             0   
4                     NaN           822     Credit Card            45   

   Driver_Ratings  Customer_Rating  Unnamed: 20  cancelled_by_customer  \
0             NaN              NaN          NaN                  False   
1             4.1              4.0          NaN                  False   
2             4.2              4.8          NaN                  False   
3             NaN              NaN          NaN                   True   
4             4.0              3.0          NaN                  False   

   cancelled_by_driver  is_incomplete  
0                 True          False  
1                False           True  
2                False           True  
3                False          False  
4                False           True  

[5 rows x 21 columns]

In [35]:
df['Incomplete_Rides'].values

array([nan, 'No', 'No', ..., 'No', 'No', 'No'], dtype=object)

In [37]:
#df = df.drop(columns=['Incomplete_Rides'])
#here true means that the ride was incomplete and false means that the ride was not incomplete
df.head()

Date      Time     Booking_ID        Booking_Status  \
0  2024-07-26 14:00:00  14:00:00  CNR7153255142    Canceled by Driver   
1  2024-07-25 22:20:00  22:20:00  CNR2940424040               Success   
2  2024-07-30 19:59:00  19:59:00  CNR2982357879               Success   
3   2024-07-22 3:15:00  03:15:00  CNR2395710036  Canceled by Customer   
4   2024-07-02 9:02:00  09:02:00  CNR1797421769               Success   

  Customer_ID Vehicle_Type Pickup_Location Drop_Location  V_TAT  C_TAT  ...  \
0   CID713523  Prime Sedan     Tumkur Road      RT Nagar    NaN    NaN  ...   
1   CID225428         Bike     Magadi Road       Varthur  203.0   30.0  ...   
2   CID270156    Prime SUV   Sahakar Nagar       Varthur  238.0  130.0  ...   
3   CID581320        eBike      HSR Layout   Vijayanagar    NaN    NaN  ...   
4   CID939555         Mini     Rajajinagar   Chamarajpet  252.0   80.0  ...   

  Incomplete_Rides_Reason Booking_Value  Payment_Method Ride_Distance  \
0                     NaN           444             NaN             0   
1                     NaN           158            Cash            13   
2                     NaN           386             UPI            40   
3                     NaN           384             NaN             0   
4                     NaN           822     Credit Card            45   

   Driver_Ratings  Customer_Rating  Unnamed: 20  cancelled_by_customer  \
0             NaN              NaN          NaN                  False   
1             4.1              4.0          NaN                  False   
2             4.2              4.8          NaN                  False   
3             NaN              NaN          NaN                   True   
4             4.0              3.0          NaN                  False   

   cancelled_by_driver  is_incomplete  
0                 True          False  
1                False           True  
2                False           True  
3                False          False  
4                False           True  

[5 rows x 21 columns]

In [39]:
df['is_incomplete'].value_counts()

is_incomplete
True     63967
False    39057
Name: count, dtype: int64

Conclusion : out of ~103024 rides
~ 63967 rides are incomplete rides 
~ 39057 rides are not completes rides

This can be true also as our dataset has is all about cancellation only

In [42]:
pd.crosstab(df['Booking_Status'], df['is_incomplete'])


is_incomplete         False  True 
Booking_Status                    
Canceled by Customer  10499      0
Canceled by Driver    18434      0
Driver Not Found      10124      0
Success                   0  63967

In [44]:
df[['is_incomplete', 'cancelled_by_customer', 'cancelled_by_driver']].sum()


is_incomplete            63967
cancelled_by_customer    10499
cancelled_by_driver      18434
dtype: int64

In [46]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103024 entries, 0 to 103023
Data columns (total 21 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Date                     103024 non-null  object 
 1   Time                     103024 non-null  object 
 2   Booking_ID               103024 non-null  object 
 3   Booking_Status           103024 non-null  object 
 4   Customer_ID              103024 non-null  object 
 5   Vehicle_Type             103024 non-null  object 
 6   Pickup_Location          103024 non-null  object 
 7   Drop_Location            103024 non-null  object 
 8   V_TAT                    63967 non-null   float64
 9   C_TAT                    63967 non-null   float64
 10  Incomplete_Rides         63967 non-null   object 
 11  Incomplete_Rides_Reason  3926 non-null    object 
 12  Booking_Value            103024 non-null  int64  
 13  Payment_Method           63967 non-null   object 
 14  Ride

In [48]:
df.isnull().sum()

Date                            0
Time                            0
Booking_ID                      0
Booking_Status                  0
Customer_ID                     0
Vehicle_Type                    0
Pickup_Location                 0
Drop_Location                   0
V_TAT                       39057
C_TAT                       39057
Incomplete_Rides            39057
Incomplete_Rides_Reason     99098
Booking_Value                   0
Payment_Method              39057
Ride_Distance                   0
Driver_Ratings              39057
Customer_Rating             39057
Unnamed: 20                103024
cancelled_by_customer           0
cancelled_by_driver             0
is_incomplete                   0
dtype: int64

In [50]:
df['Date']=pd.to_datetime(df['Date'],errors='coerce')
df['Time']=pd.to_datetime(df['Time'],format='%H:%M:%S',errors='coerce')

In [52]:
#create a new booking_datetime column
df['booking_datetime']= (
    df['Date'].dt.date.astype(str) + ' ' + df['Time'].dt.time.astype(str)
)
df['booking_datetime'] = pd.to_datetime(df['booking_datetime'], errors='coerce')

In [54]:
df[['booking_datetime']]

booking_datetime
0      2024-07-26 14:00:00
1      2024-07-25 22:20:00
2      2024-07-30 19:59:00
3      2024-07-22 03:15:00
4      2024-07-02 09:02:00
...                    ...
103019 2024-07-31 09:06:00
103020 2024-07-31 15:12:00
103021 2024-07-31 13:59:00
103022 2024-07-31 14:56:00
103023 2024-07-31 13:57:00

[103024 rows x 1 columns]

In [56]:

df = df.drop(columns=['Date', 'Time'])


In [58]:
df.head()

Booking_ID        Booking_Status Customer_ID Vehicle_Type  \
0  CNR7153255142    Canceled by Driver   CID713523  Prime Sedan   
1  CNR2940424040               Success   CID225428         Bike   
2  CNR2982357879               Success   CID270156    Prime SUV   
3  CNR2395710036  Canceled by Customer   CID581320        eBike   
4  CNR1797421769               Success   CID939555         Mini   

  Pickup_Location Drop_Location  V_TAT  C_TAT Incomplete_Rides  \
0     Tumkur Road      RT Nagar    NaN    NaN              NaN   
1     Magadi Road       Varthur  203.0   30.0               No   
2   Sahakar Nagar       Varthur  238.0  130.0               No   
3      HSR Layout   Vijayanagar    NaN    NaN              NaN   
4     Rajajinagar   Chamarajpet  252.0   80.0               No   

  Incomplete_Rides_Reason  Booking_Value Payment_Method  Ride_Distance  \
0                     NaN            444            NaN              0   
1                     NaN            158           Cash             13   
2                     NaN            386            UPI             40   
3                     NaN            384            NaN              0   
4                     NaN            822    Credit Card             45   

   Driver_Ratings  Customer_Rating  Unnamed: 20  cancelled_by_customer  \
0             NaN              NaN          NaN                  False   
1             4.1              4.0          NaN                  False   
2             4.2              4.8          NaN                  False   
3             NaN              NaN          NaN                   True   
4             4.0              3.0          NaN                  False   

   cancelled_by_driver  is_incomplete    booking_datetime  
0                 True          False 2024-07-26 14:00:00  
1                False           True 2024-07-25 22:20:00  
2                False           True 2024-07-30 19:59:00  
3                False          False 2024-07-22 03:15:00  
4                False           True 2024-07-02 09:02:00

In [60]:
new_column_order = [  # Time
    'booking_datetime',
# Identifiers
    'Booking_ID',
    'Customer_ID',
# Ride details
    'Booking_Status',
    'Vehicle_Type',
    'Pickup_Location',
    'Drop_Location',
# Operational metrics
    'V_TAT',
    'C_TAT',
    'Ride_Distance',
    'Booking_Value',
# Experience metrics
    'Payment_Method',
    'Driver_Ratings',
    'Customer_Rating',
# Cancellation / funnel flags
    'cancelled_by_customer',
    'cancelled_by_driver',
    'is_incomplete',
# Optional diagnostics
    'Incomplete_Rides_Reason'
]


In [62]:
df=df[new_column_order]

In [64]:
df.head()

booking_datetime     Booking_ID Customer_ID        Booking_Status  \
0 2024-07-26 14:00:00  CNR7153255142   CID713523    Canceled by Driver   
1 2024-07-25 22:20:00  CNR2940424040   CID225428               Success   
2 2024-07-30 19:59:00  CNR2982357879   CID270156               Success   
3 2024-07-22 03:15:00  CNR2395710036   CID581320  Canceled by Customer   
4 2024-07-02 09:02:00  CNR1797421769   CID939555               Success   

  Vehicle_Type Pickup_Location Drop_Location  V_TAT  C_TAT  Ride_Distance  \
0  Prime Sedan     Tumkur Road      RT Nagar    NaN    NaN              0   
1         Bike     Magadi Road       Varthur  203.0   30.0             13   
2    Prime SUV   Sahakar Nagar       Varthur  238.0  130.0             40   
3        eBike      HSR Layout   Vijayanagar    NaN    NaN              0   
4         Mini     Rajajinagar   Chamarajpet  252.0   80.0             45   

   Booking_Value Payment_Method  Driver_Ratings  Customer_Rating  \
0            444            NaN             NaN              NaN   
1            158           Cash             4.1              4.0   
2            386            UPI             4.2              4.8   
3            384            NaN             NaN              NaN   
4            822    Credit Card             4.0              3.0   

   cancelled_by_customer  cancelled_by_driver  is_incomplete  \
0                  False                 True          False   
1                  False                False           True   
2                  False                False           True   
3                   True                False          False   
4                  False                False           True   

  Incomplete_Rides_Reason  
0                     NaN  
1                     NaN  
2                     NaN  
3                     NaN  
4                     NaN

In [66]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103024 entries, 0 to 103023
Data columns (total 18 columns):
 #   Column                   Non-Null Count   Dtype         
---  ------                   --------------   -----         
 0   booking_datetime         103024 non-null  datetime64[ns]
 1   Booking_ID               103024 non-null  object        
 2   Customer_ID              103024 non-null  object        
 3   Booking_Status           103024 non-null  object        
 4   Vehicle_Type             103024 non-null  object        
 5   Pickup_Location          103024 non-null  object        
 6   Drop_Location            103024 non-null  object        
 7   V_TAT                    63967 non-null   float64       
 8   C_TAT                    63967 non-null   float64       
 9   Ride_Distance            103024 non-null  int64         
 10  Booking_Value            103024 non-null  int64         
 11  Payment_Method           63967 non-null   object        
 12  Driver_Ratings  

In [68]:
df.isnull().sum()

booking_datetime               0
Booking_ID                     0
Customer_ID                    0
Booking_Status                 0
Vehicle_Type                   0
Pickup_Location                0
Drop_Location                  0
V_TAT                      39057
C_TAT                      39057
Ride_Distance                  0
Booking_Value                  0
Payment_Method             39057
Driver_Ratings             39057
Customer_Rating            39057
cancelled_by_customer          0
cancelled_by_driver            0
is_incomplete                  0
Incomplete_Rides_Reason    99098
dtype: int64

In [70]:
df.head()

booking_datetime     Booking_ID Customer_ID        Booking_Status  \
0 2024-07-26 14:00:00  CNR7153255142   CID713523    Canceled by Driver   
1 2024-07-25 22:20:00  CNR2940424040   CID225428               Success   
2 2024-07-30 19:59:00  CNR2982357879   CID270156               Success   
3 2024-07-22 03:15:00  CNR2395710036   CID581320  Canceled by Customer   
4 2024-07-02 09:02:00  CNR1797421769   CID939555               Success   

  Vehicle_Type Pickup_Location Drop_Location  V_TAT  C_TAT  Ride_Distance  \
0  Prime Sedan     Tumkur Road      RT Nagar    NaN    NaN              0   
1         Bike     Magadi Road       Varthur  203.0   30.0             13   
2    Prime SUV   Sahakar Nagar       Varthur  238.0  130.0             40   
3        eBike      HSR Layout   Vijayanagar    NaN    NaN              0   
4         Mini     Rajajinagar   Chamarajpet  252.0   80.0             45   

   Booking_Value Payment_Method  Driver_Ratings  Customer_Rating  \
0            444            NaN             NaN              NaN   
1            158           Cash             4.1              4.0   
2            386            UPI             4.2              4.8   
3            384            NaN             NaN              NaN   
4            822    Credit Card             4.0              3.0   

   cancelled_by_customer  cancelled_by_driver  is_incomplete  \
0                  False                 True          False   
1                  False                False           True   
2                  False                False           True   
3                   True                False          False   
4                  False                False           True   

  Incomplete_Rides_Reason  
0                     NaN  
1                     NaN  
2                     NaN  
3                     NaN  
4                     NaN

In [85]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103024 entries, 0 to 103023
Data columns (total 19 columns):
 #   Column                   Non-Null Count   Dtype         
---  ------                   --------------   -----         
 0   booking_datetime         103024 non-null  datetime64[ns]
 1   Booking_ID               103024 non-null  object        
 2   Customer_ID              103024 non-null  object        
 3   Booking_Status           103024 non-null  object        
 4   Vehicle_Type             103024 non-null  object        
 5   Pickup_Location          103024 non-null  object        
 6   Drop_Location            103024 non-null  object        
 7   V_TAT                    63967 non-null   float64       
 8   C_TAT                    63967 non-null   float64       
 9   Ride_Distance            103024 non-null  int64         
 10  Booking_Value            103024 non-null  int64         
 11  Payment_Method           63967 non-null   object        
 12  Driver_Ratings  

In [83]:
#let's create one final conclusion column which we directly tell us whether the ride was cancelled by customer or cancelled by driver or completed or incomplete
df['ride_outcome']=np.select(
    [
        df['cancelled_by_customer'],df['cancelled_by_driver'],df['is_incomplete']
    ],['cancelled_by_customer','cancelled_by_driver','incomplete_ride'],
    default='completed_ride'
    
)

In [87]:
df['ride_outcome'].value_counts()

ride_outcome
incomplete_ride          63967
cancelled_by_driver      18434
cancelled_by_customer    10499
completed_ride           10124
Name: count, dtype: int64

In [95]:
#Since the data is for one month only lets break the booking_datetime column into 4 columns to analyse the trend
df['booking_date']=df['booking_datetime'].dt.date
df['booking_hour']=df['booking_datetime'].dt.hour
df['booking_day']=df['booking_datetime'].dt.day_name()
df['is_weekend']=df['booking_day'].isin(['Saturday','Sunday'])

In [97]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103024 entries, 0 to 103023
Data columns (total 23 columns):
 #   Column                   Non-Null Count   Dtype         
---  ------                   --------------   -----         
 0   booking_datetime         103024 non-null  datetime64[ns]
 1   Booking_ID               103024 non-null  object        
 2   Customer_ID              103024 non-null  object        
 3   Booking_Status           103024 non-null  object        
 4   Vehicle_Type             103024 non-null  object        
 5   Pickup_Location          103024 non-null  object        
 6   Drop_Location            103024 non-null  object        
 7   V_TAT                    63967 non-null   float64       
 8   C_TAT                    63967 non-null   float64       
 9   Ride_Distance            103024 non-null  int64         
 10  Booking_Value            103024 non-null  int64         
 11  Payment_Method           63967 non-null   object        
 12  Driver_Ratings  

In [101]:
df=df.drop(columns=['cancelled_by_customer','cancelled_by_driver','is_incomplete'])

In [105]:
df = df[[
    'booking_datetime',
    'booking_date',
    'booking_hour',
    'booking_day',
    'is_weekend',
    'Booking_ID',
    'Customer_ID',
    'ride_outcome',
    'Vehicle_Type',
    'Pickup_Location',
    'Drop_Location',
    'V_TAT',
    'C_TAT',
    'Ride_Distance',
    'Booking_Value',
    'Payment_Method',
    'Driver_Ratings',
    'Customer_Rating',
    'Incomplete_Rides_Reason'
]]


In [109]:
df.to_csv("rides_cleaned_july.csv",index=False)

In [111]:
!pip install sqlalchemy pymysql


[notice] A new release of pip is available: 24.3.1 -> 26.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [113]:
from sqlalchemy import create_engine
import pymysql

In [129]:
username = 'root'
password = 'root'
host = 'localhost'
database = 'ride_analysis'

engine = create_engine(f'mysql+pymysql://{'root'}:{'root'}@{'localhost'}/{'ride_analysis'}')

engine.connect()

In [131]:
df.to_sql(
    name='rides',
    con=engine,
    if_exists='replace',  # replace if table exists
    index=False
)


103024

In [139]:
df['Incomplete_Rides_Reason'].value_counts()

Incomplete_Rides_Reason
Customer Demand      1601
Vehicle Breakdown    1591
Other Issue           734
Name: count, dtype: int64